### ISSUE TITLE
CNI bridge plugin broken on containerd restart for existing containers

### ISSUE THREAD

**Issue by apostasie** [containerd/nerdctl#3355]
### Description

This is a breakout of #3352 

```
nerdctl rm -f foo
nerdctl run --net bridge -d --name foo debian sleep Inf
systemctl --user stop containerd
systemctl --user start containerd
nerdctl start foo
```

```
failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2024-08-23T14:07:01-07:00" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to allocate for range 0: 10.4.0.237 has been allocated to default-0b6b3a64bf4039828fc9443354f39009befe0240c6d0d458cefe3685ac1996e8, duplicate allocation is not allowed": unknown
```

The bridge plugin refuses to allocate an ip to this container id, since there is already one allocated...

I do believe the plugin should INSTEAD return the existing already allocated ip, but I am out of my depth here, and would welcome opinion from people more savvy with CNI networking.

### Steps to reproduce the issue

na

### Describe the results you received and expected

na

### What version of nerdctl are you using?

main

### Are you using a variant of nerdctl? (e.g., Rancher Desktop)

None

**Comment by apostasie** [comment]
Furthemore, after the above reproducer, trying to start a new unrelated container the cni plugin will error again:

```
nerdctl run -d --name bar debian sleep Inf
FATA[0000] failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2024-08-23T14:21:35-07:00" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to set bridge addr: could not set bridge's mac: invalid argument": unknown
```

**Comment by apostasie** [comment]
Finally, at the third try, it will succeed, although at that point, the namestore is all messed-up, as postStop has been called on (still) existing containers.

There is likely two problems:
- the bridge refusing to return the allocated IP (described in the original post here)
- the other error `runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2024-08-23T14:21:35-07:00" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to set bridge addr: could not set bridge's mac: invalid argument` - which suggests something wrong is being sent down to runc / cni plugin

**Comment by apostasie** [comment]
This is confirmed on a clean environment built from lima with containerd v2 + nerdctl v2.

**Comment by Nnarol** [comment]
I'm still having the issue matching the second description in this thread ("could not set bridge's mac: invalid argument") interspersed with the "allocation for IP range" variant ([Issue 665](https://github.com/containerd/nerdctl/issues/665), also believed to be fixed), in version 1.7.7, which is from September 10, while the fix is from August 24.
These kind of issues are interspersed seemingly randomly, and a different number of tries for each container makes them go away.

IP range allocation issue example:

    failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2024-11-01T13:08:49+01:00" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to allocate for range 0: 10.4.0.4 has been allocated to default-c738c11ae1a2eda42530ede520e6fe4d954a6113fd5b8574803f5140be906dde, duplicate allocation is not allowed"
    Failed to write to log, write /home/name/.local/share/nerdctl/1935db59/containers/default/c738c11ae1a2eda42530ede520e6fe4d954a6113fd5b8574803f5140be906dde/oci-hook.createRuntime.log: file already closed: unknown

Invalid argument example:

    failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2024-11-01T13:13:45+01:00" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to set bridge addr: could not set bridge's mac: invalid argument"
    Failed to write to log, write /home/name/.local/share/nerdctl/1935db59/containers/default/c738c11ae1a2eda42530ede520e6fe4d954a6113fd5b8574803f5140be906dde/oci-hook.createRuntime.log: file already closed: unknown

**Comment by AkihiroSuda** [comment]
> which is from September 10, while the fix is from August 24.

You can't just compare the date, as the Aug 24 fix is targeted for v2.0 and not backported to v1.7.

Does https://github.com/containerd/nerdctl/releases/tag/v2.0.0-rc.3 work for you?

**Comment by Nnarol** [comment]
Thank you for the quick help, I'll try!
Sorry for my failure in tracking the workflow! I wish GitHub showed some branch graphs for branches and merges.
